File: 01-Languages/Kotlin/18-Testing/tests/CalculatorTest.kt
import kotlin.test.Test
import kotlin.test.BeforeTest
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlin.test.assertFalse
import kotlin.test.assertFailsWith

class CalculatorTest {
    private lateinit var calc: Calculator // lateinit: a non-null property initialized later,
                                             // not at declaration -- Kotlin's answer to "I know
                                             // this will be set before use, but not right here"

    @BeforeTest
    fun setUp() {
        calc = Calculator() // runs before EVERY test method -- fresh instance each time
    }

    @Test
    fun addsTwoPositiveNumbers() {
        assertEquals(5, calc.add(2, 3))
    }

    @Test
    fun addsNegativeNumbers() {
        assertEquals(-5, calc.add(-2, -3))
    }

    // Table-driven style: a plain list of Triple(a, b, expected), looped over manually --
    // Kotlin/JUnit5 has @ParameterizedTest too, but a plain loop over data is equally idiomatic
    // and mirrors the array-of-tuples pattern from this repository's Rust course.
    @Test
    fun dividesCorrectly() {
        val cases = listOf(
            Triple(10.0, 2.0, 5.0),
            Triple(9.0, 3.0, 3.0),
            Triple(-6.0, 2.0, -3.0),
        )
        for ((a, b, expected) in cases) {
            assertEquals(expected, calc.divide(a, b), "divide($a, $b) failed")
        }
    }

    @Test
    fun divisionByZeroThrows() {
        val exception = assertFailsWith<IllegalArgumentException> {
            calc.divide(5.0, 0.0)
        }
        assertEquals("division by zero", exception.message)
    }

    @Test
    fun detectsPalindromes() {
        assertTrue(calc.isPalindrome("racecar"))
        assertTrue(calc.isPalindrome("A man a plan a canal Panama"))
        assertFalse(calc.isPalindrome("hello"))
        assertTrue(calc.isPalindrome(""))
    }
}

File: 01-Languages/PHP/18-Testing/tests/CalculatorTest.php
<?php
declare(strict_types=1);

namespace App\Tests;

use App\Calculator;
use PHPUnit\Framework\TestCase;
use PHPUnit\Framework\Attributes\Test;
use PHPUnit\Framework\Attributes\DataProvider;

require_once __DIR__ . "/../src/Calculator.php"; // no autoloader configured -- manual require for this lesson

final class CalculatorTest extends TestCase {
    private Calculator $calc;

    protected function setUp(): void {
        $this->calc = new Calculator(); // runs before EVERY test method -- fresh instance each time
    }

    #[Test]
    public function addsTwoPositiveNumbers(): void {
        $this->assertSame(5, $this->calc->add(2, 3));
    }

    #[Test]
    public function addsNegativeNumbers(): void {
        $this->assertSame(-5, $this->calc->add(-2, -3));
    }

    // Data provider: PHPUnit's table-driven test mechanism, analogous to Go's table tests
    // or Rust's array-of-tuples loop, but expressed as a separate method PHPUnit calls
    // automatically once per data set, reporting each case as its own named result.
    public static function divisionCases(): array {
        return [
            "ten divided by two" => [10.0, 2.0, 5.0],
            "nine divided by three" => [9.0, 3.0, 3.0],
            "negative six divided by two" => [-6.0, 2.0, -3.0],
        ];
    }

    #[Test]
    #[DataProvider('divisionCases')]
    public function dividesCorrectly(float $a, float $b, float $expected): void {
        $this->assertSame($expected, $this->calc->divide($a, $b));
    }

    #[Test]
    public function divisionByZeroThrows(): void {
        $this->expectException(\InvalidArgumentException::class);
        $this->expectExceptionMessage("division by zero");
        $this->calc->divide(5.0, 0.0);
    }

    #[Test]
    #[DataProvider('palindromeCases')]
    public function detectsPalindromes(string $input, bool $expected): void {
        $this->assertSame($expected, $this->calc->isPalindrome($input));
    }

    public static function palindromeCases(): array {
        return [
            "racecar" => ["racecar", true],
            "phrase with spaces and case" => ["A man a plan a canal Panama", true],
            "not a palindrome" => ["hello", false],
            "empty string" => ["", true],
        ];
    }
}
